### PR TITLE
Adding config option to run docker commands in WSL

### PR DESCRIPTION
**Pull request by dmchurch**
This adds a couple new configuration items that let Container Tools run the Docker commands in a WSL container. There's probably some amount of improvement that can be done to autodetect whether that should be enabled rather than requiring the user to find the config option, but I figure that can be future work.

Also, I'm fully expecting this will come second in line behind #165, whenever that lands, and that's fine; I might as well get the concept out there now, even if I have to refactor it later.

### CODE REVIEW

**Comment by dmchurch** [comment]
@microsoft-github-policy-service agree

**Comment by dmchurch** [comment]
Yup, there we go, those are the conflicts I was expecting. I'll work on updating my PR to adjust, but @bwateratmsft, if you have a chance to look at this definitely lmk if there's anything architectural that needs changing here 🙂

File: src/runtimes/ExecutionEnvironmentManager.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        // Register an event to watch for changes to config, reconfigure if needed
+        registerEvent('vscode-containers.command.changed', vscode.workspace.onDidChangeConfiguration, (actionContext: IActionContext, e: vscode.ConfigurationChangeEvent) => {
+            actionContext.telemetry.suppressAll = true;
+            actionContext.errorHandling.suppressDisplay = true;
+
+            if (e.affectsConfiguration(`${configPrefix}.executeInWSL`) || e.affectsConfiguration(`${configPrefix}.executeInWSLDistro`)) {
+                this.reconfigure();
+            }
+        });
+    }
```
[the remark this answers is not in the document]
Comment (dmchurch) [reply src/runtimes/ExecutionEnvironmentManager.ts:55]: Aha! That explains a great deal, and I'm extremely enthusiastic about not having to go through that effort! Runtime lookup is indeed the way to go here. :100:

File: src/runtimes/runners/runWithDefaults.ts [hunk 2]
```diff
@@ -149,6 +149,11 @@ class DefaultEnvStreamCommandRunnerFactory<TOptions extends DefaultEnvStreamComm
         this.errAccumulator = errAccumulator;
     }
 
+    protected override getCommandAndArgs(commandResponse: CommandResponseBase): { command: string; args: CommandLineArgs; } {
+        // This needs to remain unquoted as it's being passed to child_process.spawn
+        return ext.executionManger.adjustExecutionArguments(super.getCommandAndArgs(commandResponse), {shouldQuote: false});
+    }
```
[the remark this answers is not in the document]
Comment (dmchurch) [reply src/runtimes/runners/runWithDefaults.ts:155]: That was my first impulse as well (and was, in fact, my inspiration for the PR in the first place)! I saw that the WSCRF class was there but wasn't being used, so my first change to DESCRO was to change it from an inheritance to a composition so I could do exactly that.

Then I ran into issues with shell quoting - WSCRF ends up applying quotes to the distro argument, but since that gets passed directly to `wsl.exe` in a `child_process.spawn` call, the quotation marks end up as part of the value and the command fails (because while `Ubuntu` is a valid distro on my system, `"Ubuntu"` is not). The task runner doesn't suffer from this problem, but the `runWithDefaults` and `execAsync` runners both have that issue.

Currently I'm expanding the role of the `ExecutionEnvironmentManager` a bit; now, it can massage not just the command and args, but also the spawn options as well. That's because my current hurdle is `getServiceSubsets`, which just puts a workspace-supplied command line into `command` and leaves `args` empty, [requiring the use of `allowUnsafeExecutablePath`](https://github.com/microsoft/vscode-containers/blob/main/src/commands/compose/getComposeSubsetList.ts#L244). aUEP causes problems for the WSL case, though, because it causes `spawnStreamAsync` to skip some of its own processing. So, the executionManager now has to check for and disable the `allowUnsafeExecutablePath`, if and only if this system is in WSL mode.

I'll keep an eye on the architecture, though, because I think we probably still haven't seen its final form; among other things, I'm thinking that there's going to be a need to support multiple independent execution managers in the same way as we have independent shell-quoting providers, even running in the same host. For example, you might have some commands that want to get run on the local system vs some that want to be run on the Docker host or that sort of thing; I could even see having an implementation that lets you access the docker commands over an SSH link! Perhaps one that lets you run the docker commands with a sudo wrapper too, who knows.

That, of course, brings up a related complication that we're not totally free of even now, since we have multiple different potential shells involved; there might end up being a need to encode a single command line using the quoting styles from two separate shells on two different platforms. Thankfully the command line arguments involved don't tend to use much in the way of weird characters or whitespace, but it's certainly something to keep in mind!

File: package.nls.json [hunk 3]
```diff
@@ -204,6 +204,8 @@
     "vscode-containers.config.containers.enableComposeLanguageService": "Whether or not to enable the Compose Language Service. Changing requires a restart to take effect.",
     "vscode-containers.config.containers.containerClient": "Which container client to use. If not specified, Docker will be used. Changing requires a restart to take effect.",
     "vscode-containers.config.containers.orchestratorClient": "Which container orchestrator client to use. If not specified, Docker Compose will be used. Changing requires a restart to take effect.",
+    "vscode-containers.config.containers.executeInWSL": "Controls whether container and compose actions should be executed in WSL. This setting has no effect on other platforms than Windows.",
+    "vscode-containers.config.containers.executeInWSLDistro": "WSL distribution to use when executing container actions. The default is to use the default WSL distro (see `wsl -l`). This setting has no effect when `#vscode-containers.config.containers.executeInWSL#` is not enabled or on other platforms than Windows.",
```
[the remark this answers is not in the document]
Comment (dmchurch) [reply package.nls.json:208]: Aha, okay, I'll give that a shot!

File: src/runtimes/runners/runWithDefaults.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bwateratmsft) [reply src/runtimes/runners/runWithDefaults.ts:155]: > That's because my current hurdle is getServiceSubsets, which just puts a workspace-supplied command line into command and leaves args empty, [requiring the use of allowUnsafeExecutablePath](https://github.com/microsoft/vscode-containers/blob/main/src/commands/compose/getComposeSubsetList.ts#L244)

Oh my goodness I know!! That intersection of command customization and compose subsetting is a pain for us too!

File: package.json [hunk 4]
```diff
@@ -2328,6 +2328,18 @@
                         "Docker Compose",
                         "Podman Compose"
                     ]
+                },
+                "containers.executeInWSL": {
```
[the remark this answers is not in the document]
Comment (bwateratmsft) [reply package.json:2332]: Nice!

**Comment by dmchurch** [comment]
Y'know, the more I look at this, the more I'm wondering if the base concept of the `ExecutionEnvironmentManager`—of the middleware which gets to dictate how a process should get run, right before it gets spawned, but not the parts that talk to the extension configuration—wants to live in vscode-processutils itself. Do you want me to shift the machinery over to a PR on that repo instead of keeping it in here?

**Comment by danegsta** [comment]
> Fyi, I'm still not confident in my conclusions, but I'm beginning to suspect that @danegsta was at least somewhat mistaken about the behavior of `windowsVerbatimArguments` in the [comment in `NoShell`](https://github.com/microsoft/vscode-docker-extensibility/blob/main/packages/vscode-processutils/src/utils/Shell.ts#L177-L187). That code is what's responsible for `NoShell` putting quotes around the distro name and causing WSL to reject it, so it's definitely the case that Node _isn't_ stripping the quotes out before putting them in wsl.exe's argv array.
> 
> I suspect—though I can't be certain, of course—that the issues here stem from the fact that when NoShell was introduced, the only thing it was being used for was [a direct, non-shell execution of the cmd.exe binary](https://github.com/microsoft/vscode-docker-extensibility/blob/d1b355ef111fb4fe9216d5a271ba0e3765299780/packages/container-runtimes/src/clients/DockerClientBase/DockerClientBase.ts#L1560-L1566) in [microsoft/vscode-docker-extensibility#152](https://github.com/microsoft/vscode-docker-extensibility/pull/152). In other words, while this was in fact a situation where the spawn call needed to have both `windowsVerbatimArguments: true` _and_ `shell: false`, that wasn't because some combination of Windows + Node requires double-quotes in some non-shell executions. My guess is it was because (a) cmd.exe is a little bit pathological about how it does command-line parameter interpretation, and (b) that particular call in `DockerClientBase` is mixing string-per-arguments from L1561-L1564 with the single space-delimited string of L1565.
> 
> If I'm right, then that means that the quote-adding behavior of NoShell is, in fact, an outright bug, which means this PR _definitely_ needs to extend over to the other repository 😅

`windowsVerbatimArguments` is a weird flag; when set, it treats all arguments as "raw" arguments (as though you'd typed them out verbatim on the command line), but Windows also does some basic quoting logic even when a command isn't run via a shell. So when using `windowsVerbatimArguments`, `["--source", "C:\\some\\path with\\spaces"]` becomes three separate arguments to a process (`--source`, `C:\some\path`, and `with\spaces`). While `["--source", "\"C:\\some\\path with\\spaces\""]` becomes (`--source`, `C:\some\path with\spaces`).

We don't strictly _have_ to use `windowsVerbatimArguments` when running without a shell (in some ways the resulting logic without it makes a lot more sense as each individual argument in the array becomes a discrete argument out to the running process, spaces or not). The main headache is that when using a shell on Windows, node forces on `windowsVerbatimArguments`, so we have to manage different assumptions around quoting behavior with and without a shell. I'm more than happy to stop using it for NoShell if we can make it work for all command scenarios (spaces in child arguments in things like environment variables, labels, or volume paths).

In the case of WSL, you may want to actually use `--exec` instead of `--` to avoid running the command via shell in the distro. Ultimately, the main headache is that you actually have two effectively nested quoting/escaping behaviors to worry about (`wsl -d <distro> --exec <command>` is processed by Windows first and foremost (either without a shell or via a shell in the terminal) before `<command>` is evaluated in the context of the WSL distro. Either way, <command> likely needs to be treated as raw arguments rather than depending on the shell (or NoShell) implementation to get the quoting correct for the WSL distro (see `withVerbatimArg` from `commandLineBuilder`).

I think there are definitely some changes that could be made in [commandLineBuilder.ts](https://github.com/microsoft/vscode-docker-extensibility/blob/main/packages/vscode-processutils/src/utils/commandLineBuilder.ts) and [Shell.ts](https://github.com/microsoft/vscode-docker-extensibility/blob/main/packages/vscode-processutils/src/utils/Shell.ts) in https://github.com/microsoft/vscode-docker-extensibility that would make this problem more straightforward.

**Comment by dmchurch** [comment]
> Windows also does some basic quoting logic even when a command isn't run via a shell

I'm not entirely certain that's correct, at least not in this particular context. The behavior I'm seeing here strongly points to the idea that Windows is _not_ doing any quote processing, or that if it is, Node is counteracting it. I suspect the latter, and that what the `windowsVerbatimArguments` _does_ is instruct Node to perform its own escaping to counteract Windows's command-line argument processing. If that's the case, and I suspect it is, then the NoShell code is duplicating work and we should remove it in favor of using the builtin Node functionality.

**Comment by danegsta** [comment]
> > Windows also does some basic quoting logic even when a command isn't run via a shell
> 
> I'm not entirely certain that's correct, at least not in this particular context. The behavior I'm seeing here strongly points to the idea that Windows is _not_ doing any quote processing, or that if it is, Node is counteracting it. I suspect the latter, and that what the `windowsVerbatimArguments` _does_ is instruct Node to perform its own escaping to counteract Windows's command-line argument processing. If that's the case, and I suspect it is, then the NoShell code is duplicating work and we should remove it in favor of using the builtin Node functionality.

If you run a command with shell on Windows (at least for Cmd), `windowsVerbatimArguments` is forcibly enabled. If you run without shell (which we normally do outside of user facing commands that are run in a terminal), `windowsVerbatimArguments` is optional. All `windowsVerbatimArguments` does is disable the quoting/escaping logic node would normally do for command line arguments (and only on Windows).

The underlying win32 APIs for creating a process are pretty basic; all the arguments passed to [CreateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa) are sent as a single concatenated string. Technically its up to each individual Windows program to parse the raw command line. A given programming language on Windows either automatically includes a bit of extra code to do this parsing (C++ or node.js, for example), includes command line parsing in its standard library (Go), or leaves it up to the user. This means that any given program can interpret the individual arguments differently.

Our current implementation has mostly been focused on getting the correct behavior from Go applications (since both Docker and Podman are written in Go), but it sounds like wsl.exe parses arguments differently and doesn't strip quotes in the same way. When setting `windowsVerbatimArguments: false`, node is fully taking over quoting and escaping the arguments, but it's doing so in the specific way that node also parses arguments and that may not be exactly the way wsl.exe or a Go application expects to parse quoted arguments.

I'm fully on board with disabling `windowsVerbatimArguments` for our `NoShell` scenario if we can show it isn't breaking anything when dealing with edge cases with complex Docker/Podman arguments containing spaces. I've thrown together a simple 12 line Go app that echos back all the args it parsed and have a test case I'm working on that passes NoShell args and checks the output to make sure they match. With that we should be able to at confirm Go is recognizing the arguments the same way after the change.

**Comment by dmchurch** [comment]
> I've thrown together a simple 12 line Go app that echos back all the args it parsed and have a test case I'm working on that passes NoShell args and checks the output to make sure they match. With that we should be able to at confirm Go is recognizing the arguments the same way after the change.

Ha, I've been doing a very similar thing (though my tester is in C++, rather than Go):

image

...which makes me all the more confused because it's suggesting basically opposite behavior to what I was seeing from the extension, so I'm gonna have to drill down further to figure out exactly what it was I was misinterpreting. (Yes, my tests are showing me the same thing you were trying to explain to me. Some people just won't grasp something until they try it themself, but I am quite aware that I'm one of those people, so I try to be proactive in trying things myself!)

(Thanks for the refresher on the Windows process-call ABI btw! It's been a long time since I've done work that low-level on a win32 system, and my head always gets a bit turned around when I have to switch between high-level and low-level contexts too rapidly 😂)

Anyway, yes, I'm onboard with trying to make the non-shell calls be `windowsVerbatimArguments:false` calls, but that's going to be at least somewhat tricky to ensure that it's done correctly, in part because we _do_ have use places where full command-lines are being passed as a single string. Please correct me if I'm wrong, but my interpretation is this:

### For non-shell calls, the only way for a single `String` in the `cp.spawn()` arguments array to be interpreted as multiple arguments by an `int main()`-conforming child process is if the `windowsVerbatimArguments` option is `true`.

Continuing to think through it, that's why setting `shell:true` forces wVA to be true, because when you call spawn that way, the whole point is that you want the spaces in your string to be delimited. I think it's probably slightly more accurate from the perspective of the calling code, though, to say that setting `shell:true` causes wVA to be _irrelevant_. All the arguments you pass to the spawn function are already getting indirected by turning it into a `cmd.exe /c` command line, and it's only Node's internally-constructed arguments list that `wVA:true` applies to.

----

All the above notwithstanding, it's definitely still the case that with the code in my PR, allowing `NoShell` to wrap the distro argument in quotes caused `wsl.exe` to fail to recognize the distribution, but inhibiting the quotes made it work. I'll keep digging; my next plan is to add the `wslPath` option to allow the user to choose an executable besides `wsl.exe`, so I can get some better logging.
